File: deploy/backend/spring-gateway/src/test/java/com/yojnasetu/gateway/credit/EmiCalculatorTest.java
package com.yojnasetu.gateway.credit;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Worked examples deliberately mirror frontend/src/lib/emiCalculator.test.js —
 * the two implementations must agree, and testing them against the same numbers
 * is what makes a divergence show up as a failure rather than as two plausible
 * figures on two screens.
 */
class EmiCalculatorTest {

    // Micro Finance Scheme's real seeded parameters.
    private static final long PRINCIPAL = 140_000;
    private static final double RATE = 6.5;
    private static final int TENURE = 36;
    private static final int MORATORIUM = 3;

    private static void near(double expected, double actual) {
        near(expected, actual, 0.01);
    }

    private static void near(double expected, double actual, double tolerance) {
        assertTrue(Math.abs(expected - actual) < tolerance,
                () -> "expected ~" + expected + " but was " + actual);
    }

    @Test
    void matchesTheStandardFormulaWithoutAMoratorium() {
        EmiPlan plan = EmiCalculator.calculate(100_000, 12, 12, 0, MoratoriumMode.CAPITALISE);
        double r = 0.01;
        double f = Math.pow(1 + r, 12);
        near((100_000 * r * f) / (f - 1), plan.emi());
        assertEquals(12, plan.schedule().size());
    }

    @Test
    void treatsAZeroRateAsAStraightPrincipalSplit() {
        EmiPlan plan = EmiCalculator.calculate(120_000, 0, 12, 6, MoratoriumMode.CAPITALISE);
        near(10_000, plan.emi());
        near(0, plan.totalInterest());
        near(0, plan.moratoriumInterest());
    }

    @Test
    void capitalisesAccruedInterestIntoThePrincipal() {
        EmiPlan plan = EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, MORATORIUM, MoratoriumMode.CAPITALISE);
        near(142_287.35, plan.financedPrincipal(), 0.05);
        near(2_287.35, plan.moratoriumInterest(), 0.05);
        near(0, plan.moratoriumPayment());
    }

    @Test
    void aMoratoriumAlwaysCostsMoreThanNone() {
        // The regression that shipped in the frontend: a moratorium was shown in
        // the UI but never entered the arithmetic, so this came out equal.
        EmiPlan with = EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, MORATORIUM, MoratoriumMode.CAPITALISE);
        EmiPlan without = EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, 0, MoratoriumMode.CAPITALISE);
        assertTrue(with.totalInterest() > without.totalInterest());
        assertTrue(with.emi() > without.emi());
    }

    @Test
    void servicingInterestLeavesThePrincipalAloneAndCostsLessOverall() {
        EmiPlan serviced = EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, MORATORIUM, MoratoriumMode.SERVICE_INTEREST);
        EmiPlan capitalised = EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, MORATORIUM, MoratoriumMode.CAPITALISE);

        near(PRINCIPAL, serviced.financedPrincipal());
        near(PRINCIPAL * (RATE / 12 / 100) * MORATORIUM, serviced.moratoriumPayment());
        near(EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, 0, MoratoriumMode.CAPITALISE).emi(), serviced.emi());
        assertTrue(serviced.totalPayment() < capitalised.totalPayment());
    }

    @ParameterizedTest
    @EnumSource(MoratoriumMode.class)
    void scheduleSpansTheWholeLoanLifeWithAbsoluteMonthNumbers(MoratoriumMode mode) {
        List<EmiPlan.ScheduleRow> rows =
                EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, MORATORIUM, mode).schedule();

        assertEquals(MORATORIUM + TENURE, rows.size());
        for (int i = 0; i < rows.size(); i++) {
            assertEquals(i + 1, rows.get(i).month());
            EmiPlan.Phase expected = i < MORATORIUM ? EmiPlan.Phase.MORATORIUM : EmiPlan.Phase.REPAYMENT;
            assertEquals(expected, rows.get(i).phase());
        }
    }

    @ParameterizedTest
    @EnumSource(MoratoriumMode.class)
    void amortisesToZeroAndReconcilesInterest(MoratoriumMode mode) {
        EmiPlan plan = EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, MORATORIUM, mode);

        near(0, plan.schedule().get(plan.schedule().size() - 1).balance());

        // Row-level interest must add up to the headline figure — an identity
        // that holds in both modes and catches most arithmetic drift.
        double summed = plan.schedule().stream()
                .mapToDouble(EmiPlan.ScheduleRow::interestComponent)
                .sum();
        near(plan.totalInterest(), summed, 0.02);
    }

    @Test
    void noEmiIsDueDuringACapitalisedMoratorium() {
        EmiPlan plan = EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, MORATORIUM, MoratoriumMode.CAPITALISE);
        plan.schedule().stream()
                .filter(r -> r.phase() == EmiPlan.Phase.MORATORIUM)
                .forEach(r -> {
                    near(0, r.emi());
                    near(0, r.principalComponent());
                    assertTrue(r.interestComponent() > 0);
                });
    }

    @Test
    void educationLoanAccruesTwelveMonthsBeforeTheFirstEmi() {
        EmiPlan plan = EmiCalculator.calculate(2_000_000, 6.5, 120, 12, MoratoriumMode.CAPITALISE);
        near(2_000_000 * Math.pow(1 + 0.065 / 12, 12), plan.financedPrincipal(), 0.05);
        assertTrue(plan.moratoriumInterest() > 130_000);
        assertEquals(12, plan.schedule().stream()
                .filter(r -> r.phase() == EmiPlan.Phase.MORATORIUM).count());
    }

    @Test
    void defaultsToCapitalisingWhenNoModeIsGiven() {
        EmiPlan explicit = EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, MORATORIUM, MoratoriumMode.CAPITALISE);
        EmiPlan defaulted = EmiCalculator.calculate(PRINCIPAL, RATE, TENURE, MORATORIUM, null);
        near(explicit.emi(), defaulted.emi());
        assertEquals(MoratoriumMode.CAPITALISE, defaulted.moratoriumMode());
    }
}

File: frontend/src/lib/emiCalculator.test.js
import { describe, it, expect } from 'vitest'
import {
  calculateEmi,
  formatInr,
  MORATORIUM_CAPITALISE,
  MORATORIUM_SERVICE_INTEREST,
} from './emiCalculator'

// The Micro Finance Scheme's real parameters, as served by the backend
// catalogue (GET /api/v2/sih/credit/products).
//
// The principal here was 140000, copied from the old hardcoded
// lib/nsfdcSchemes.js — but ₹1.40 lakh is the scheme's PROJECT-COST ceiling,
// not its loan cap, which is ₹1.25 lakh. Only the EMI maths is under test so
// the assertions held either way, but leaving the wrong number labelled "the
// Micro Finance Scheme's real parameters" is how that figure kept spreading.
const MFS = { principal: 125000, rate: 6.5, tenure: 36, moratorium: 3 }

const near = (a, b, tolerance = 0.01) => expect(Math.abs(a - b)).toBeLessThan(tolerance)

describe('calculateEmi — no moratorium', () => {
  it('matches the standard reducing-balance formula', () => {
    const { emi } = calculateEmi(100000, 12, 12)
    // P·r·(1+r)^n / ((1+r)^n − 1) with r = 0.01
    const r = 0.01
    const f = Math.pow(1 + r, 12)
    near(emi, (100000 * r * f) / (f - 1))
  })

  it('handles a zero interest rate as a straight principal split', () => {
    const { emi, totalInterest, totalPayment } = calculateEmi(120000, 0, 12)
    near(emi, 10000)
    near(totalInterest, 0)
    near(totalPayment, 120000)
  })

  it('is unchanged when called with the old three-argument signature', () => {
    // Guards the existing callers: omitting options must mean "no moratorium".
    const threeArg = calculateEmi(140000, 6.5, 36)
    const explicit = calculateEmi(140000, 6.5, 36, { moratoriumMonths: 0 })
    near(threeArg.emi, explicit.emi)
    near(threeArg.totalPayment, explicit.totalPayment)
    expect(threeArg.schedule).toHaveLength(36)
  })
})

describe('calculateEmi — capitalised moratorium', () => {
  const result = calculateEmi(MFS.principal, MFS.rate, MFS.tenure, {
    moratoriumMonths: MFS.moratorium,
    moratoriumMode: MORATORIUM_CAPITALISE,
  })

  it('capitalises accrued interest into the principal', () => {
    // 125000 × (1 + 0.065/12)^3
    near(result.financedPrincipal, 127042.27, 0.05)
    near(result.moratoriumInterest, 2042.27, 0.05)
  })

  it('collects nothing from the citizen during the moratorium', () => {
    near(result.moratoriumPayment, 0)
    for (const row of result.schedule.filter((r) => r.phase === 'moratorium')) {
      near(row.emi, 0)
      near(row.principalComponent, 0)
      expect(row.interestComponent).toBeGreaterThan(0)
    }
  })

  it('costs strictly more than the same loan with no moratorium', () => {
    // This is the regression the old implementation shipped: it returned the
    // no-moratorium figure while the UI claimed a moratorium was in effect.
    const noMoratorium = calculateEmi(MFS.principal, MFS.rate, MFS.tenure)
    expect(result.totalInterest).toBeGreaterThan(noMoratorium.totalInterest)
    expect(result.emi).toBeGreaterThan(noMoratorium.emi)
  })
})

describe('calculateEmi — interest-servicing moratorium', () => {
  const result = calculateEmi(MFS.principal, MFS.rate, MFS.tenure, {
    moratoriumMonths: MFS.moratorium,
    moratoriumMode: MORATORIUM_SERVICE_INTEREST,
  })

  it('leaves the principal untouched', () => {
    near(result.financedPrincipal, MFS.principal)
  })

  it('charges simple interest for each moratorium month', () => {
    const monthly = MFS.principal * (MFS.rate / 12 / 100)
    near(result.moratoriumInterest, monthly * MFS.moratorium)
    near(result.moratoriumPayment, monthly * MFS.moratorium)
    for (const row of result.schedule.filter((r) => r.phase === 'moratorium')) {
      near(row.emi, monthly)
      near(row.balance, MFS.principal)
    }
  })

  it('keeps the EMI identical to the no-moratorium loan', () => {
    const noMoratorium = calculateEmi(MFS.principal, MFS.rate, MFS.tenure)
    near(result.emi, noMoratorium.emi)
  })

  it('costs less overall than capitalising', () => {
    const capitalised = calculateEmi(MFS.principal, MFS.rate, MFS.tenure, {
      moratoriumMonths: MFS.moratorium,
      moratoriumMode: MORATORIUM_CAPITALISE,
    })
    expect(result.totalPayment).toBeLessThan(capitalised.totalPayment)
  })
})

describe('schedule integrity', () => {
  for (const mode of [MORATORIUM_CAPITALISE, MORATORIUM_SERVICE_INTEREST]) {
    describe(mode, () => {
      const result = calculateEmi(MFS.principal, MFS.rate, MFS.tenure, {
        moratoriumMonths: MFS.moratorium,
        moratoriumMode: mode,
      })

      it('spans the whole loan life with absolute month numbers', () => {
        expect(result.schedule).toHaveLength(MFS.moratorium + MFS.tenure)
        expect(result.schedule.map((r) => r.month)).toEqual(
          Array.from({ length: MFS.moratorium + MFS.tenure }, (_, i) => i + 1),
        )
        // The moratorium is the opening stretch, not an afterthought.
        expect(result.schedule.slice(0, MFS.moratorium).every((r) => r.phase === 'moratorium')).toBe(true)
        expect(result.schedule.slice(MFS.moratorium).every((r) => r.phase === 'repayment')).toBe(true)
      })

      it('amortises to a zero balance', () => {
        near(result.schedule.at(-1).balance, 0)
      })

      it('reconciles row-level interest against the reported total', () => {
        const summed = result.schedule.reduce((acc, r) => acc + r.interestComponent, 0)
        near(summed, result.totalInterest, 0.02)
      })
    })
  }
})

describe('education loan — the longest moratorium in the catalogue', () => {
  it('accrues twelve months of interest before the first EMI', () => {
    const { financedPrincipal, moratoriumInterest, schedule } = calculateEmi(2000000, 6.5, 120, {
      moratoriumMonths: 12,
      moratoriumMode: MORATORIUM_CAPITALISE,
    })
    near(financedPrincipal, 2000000 * Math.pow(1 + 0.065 / 12, 12), 0.05)
    expect(moratoriumInterest).toBeGreaterThan(130000)
    expect(schedule.filter((r) => r.phase === 'moratorium')).toHaveLength(12)
  })
})

describe('formatInr', () => {
  it('renders whole rupees in the Indian digit grouping', () => {
    expect(formatInr(140000)).toBe('₹1,40,000')
    expect(formatInr(2287.35)).toBe('₹2,287')
  })
})
